internal/campaigns/run_steps.go

```diff
@@ -271,3 +331,171 @@ func (e stepFailedErr) SingleLineError() string {
 
 	return strings.Split(out, "\n")[0]
 }
+
+func parseStepRun(run string, stepCtx *StepContext) (*template.Template, error) {
+	return template.New("step-run").Delims("${{", "}}").Funcs(stepCtx.ToFuncMap()).Parse(run)
+}
+
+func parseStepFiles(files map[string]interface{}, stepCtx *StepContext) (map[string]string, error) {
+	containerFiles := make(map[string]string, len(files))
+
+	fnMap := stepCtx.ToFuncMap()
+
+	for fileName, fileRaw := range files {
+		// We treat the file contents as a template and render it
+		// into a buffer that we then mount into the code host.
+		var out bytes.Buffer
+
+		tmpl, err := template.New(fileName).Delims("${{", "}}").Funcs(fnMap).Parse(fileRaw.(string))
+		if err != nil {
+			return containerFiles, err
+		}
+
+		if err := tmpl.Execute(&out, stepCtx); err != nil {
+			return containerFiles, err
+		}
+
+		containerFiles[fileName] = out.String()
+	}
+
+	return containerFiles, nil
+}
+
+func parseStepEnv(env map[string]string, stepCtx *StepContext) (map[string]string, error) {
+	parsedEnv := make(map[string]string, len(env))
+
+	fnMap := stepCtx.ToFuncMap()
+
+	for k, v := range env {
+		// We treat the file contents as a template and render it
+		// into a buffer that we then mount into the code host.
+		var out bytes.Buffer
+
+		tmpl, err := template.New(k).Delims("${{", "}}").Funcs(fnMap).Parse(v)
+		if err != nil {
+			return parsedEnv, err
+		}
+
+		if err := tmpl.Execute(&out, stepCtx); err != nil {
+			return parsedEnv, err
+		}
+
+		parsedEnv[k] = out.String()
+	}
+
+	return parsedEnv, nil
+}
+
+// StepContext represents the contextual information available when executing a
+// step that's defined in a campaign spec.
+type StepContext struct {
+	PreviousStep StepResult
+	Repository   graphql.Repository
+}
+
+// ToFuncMap returns a template.FuncMap to access fields on the StepContext in a
+// text/template.
+func (stepCtx *StepContext) ToFuncMap() template.FuncMap {
+	return template.FuncMap{
+		"join": func(list []string, sep string) string {
+			return strings.Join(list, sep)
+		},
+		"previous_step": func() map[string]interface{} {
+			result := map[string]interface{}{
+				"modified_files": stepCtx.PreviousStep.ModifiedFiles(),
+				"added_files":    stepCtx.PreviousStep.AddedFiles(),
+				"deleted_files":  stepCtx.PreviousStep.DeletedFiles(),
+			}
+
+			if stepCtx.PreviousStep.Stdout != nil {
```

Comment:
What happens if I use `${{ previous_step }}` in the first step? Would this be nil and panic?

Reply:
No. Say you have this in your spec:

```yaml
on:
  - repositoriesMatchingQuery: repo:automation-testing

steps:
  - run: |
      cat ${{ previous_step.modified_files }} >> previous_step_modified_files.txt &&\
        cat ${{ repository.search_result_paths }} >> repository_search_result_paths.txt
    container: alpine:3
```

that would produce two empty files. No panic, no errors.